### ISSUE TITLE
Problem building 0.5 with MKL

### ISSUE THREAD

**Issue by lruthotto** [JuliaLang/julia#18167]
I'm having trouble compiling the 0.5 release branch with MKL on Mac OS X 10.10.5. 

My `Make.user` file looks like this:

```
USE_INTEL_MKL = 1
USE_INTEL_MKL_FFT = 1
```

I've followed the steps here (https://software.intel.com/en-us/articles/julia-with-intel-mkl-for-improved-performance) and upon `make` the output I'm getting is this: 

```
ld: warning: directory not found for option '-L/lib/ia32'
ld: warning: directory not found for option '-L/lib/ia32'
ld: library not found for -lmkl_rt
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [~/Software/julia-0.5/usr/lib/libspqr.dylib] Error 1
make: *** [julia-deps] Error 2
```

I'm not sure this is an issue with my MKL setup as I have no trouble compiling the 0.4 release branch.

**Comment by ViralBShah** [comment]
I thought our MKL support was only ever tested on linux.

**Comment by vtjnash** [comment]
From the error message, it looks like you skipped the step of setting MKLROOT to the directory where you installed MKL. But since it worked for you on v0.4, maybe this variable changed or bitrotted?

**Comment by lruthotto** [comment]
Okay, good to know that this is not supposed to work, but I was pretty happy with MKL under 0.4. 

I don't think MKLROOT is the issue:
```
$ echo $MKLROOT
/opt/intel//compilers_and_libraries_2016.3.170/mac/mkl
ls $MKLROOT
benchmarks	examples	interfaces	samples
bin		include		lib		tools
```

**Comment by vtjnash** [comment]
Possibly a silly question, but is it also 'export'ed?

**Comment by lruthotto** [comment]
seems so. 

``` 
export MKLROOT
make
```

resulted in the same error. Thanks for looking into this!!

**Comment by vtjnash** [comment]
Did you remember to distclean suitesparse after exporting the variable?

What is 'make print-MKLLIB'?

**Comment by lruthotto** [comment]
What does disclean suitesparse mean? Can you send me a command?

```
$ make print-MKLLIB
MKLLIB=/opt/intel//compilers_and_libraries_2016.3.170/mac/mkl/lib/intel64
```

**Comment by lruthotto** [comment]
Thanks, Tony. Tried it, but unfortunately it didn't resolve the problem.

**Comment by mouryarahul** [comment]
@lruthotto Hi, I've recently build Julia-0.5 using Intel icc (ICC) 16.0.3 20160415 without any problems. Here is the VersionInfo: 
Julia Version 0.5.1-pre+2
Commit f0d40ec* (2016-09-20 03:34 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-2670QM CPU @ 2.20GHz
  WORD_SIZE: 64
  BLAS: libmkl_rt
  LAPACK: libmkl_rt
  LIBM: libimf
  LLVM: libLLVM-3.7.1 (ORCJIT, sandybridge)

**Comment by KristofferC** [comment]
I have also build julia 0.5 with MKL without problems. @lruthotto did you manage to figure this out?

**Comment by lruthotto** [comment]
kind of. we used the intel compilers for now and upgraded the Ubuntu version. The speedup was not impressive though (fftn in MATLAB was still faster by 2x factor than fft! in Julia+MKL) so there might be still something wrong with my setup.

**Comment by KristofferC** [comment]
Please comment if there is still something to do here.

**Comment by fasiha** [comment]
@KristofferC was this with icc or with clang? I’m trying to install with clang and running into a similar error.

- I installed Intel MKL 2017 on macOS.
- I do `source /opt/intel/mkl/bin/mklvars.sh intel64 ilp64`.
- I clone Julia and checkout the 0.5.2 tag, then in Make.inc I modify:
```make
USE_INTEL_MKL = 1
USE_INTEL_MKL_FFT = 1
```
- I run `make -j4` and `JULIA usr/lib/julia/sys.o` hits the error: `LoadError("sysimg.jl",298,LoadError("linalg/linalg.jl",225,LoadError("linalg/blas.jl",79,ErrorException("error compiling vendor: could not load library \"libmkl_rt\"\ndlopen(libmkl_rt.dylib, 1): image not found"))))`.

Ok, so I dig into the Make.inc and notice starting at line 971:
```make
ifeq ($(USE_INTEL_MKL), 1)
ifeq ($(USE_BLAS64), 1)
export MKL_INTERFACE_LAYER := ILP64
MKLLIB := $(MKLROOT)/lib/intel64 ########################### <- wat???
else
MKLLIB := $(MKLROOT)/lib/ia32
endif
USE_SYSTEM_BLAS:=1
USE_SYSTEM_LAPACK:=1
ifeq ($(OS), WINNT)
LIBBLASNAME := mkl_rt
LIBLAPACKNAME := mkl_rt
# work around libtool issue with arpack
MKL_LDFLAGS := -L$(MKLLIB) -Wl,-lmkl_rt
else
LIBBLASNAME := libmkl_rt
LIBLAPACKNAME := libmkl_rt
MKL_LDFLAGS := -L$(MKLLIB) -lmkl_rt
endif
ifneq ($(strip $(MKLLIB)),)
  ifeq ($(OS), Linux)
    RPATH_MKL := -Wl,-rpath,$(MKLLIB)
    RPATH     += $(RPATH_MKL)
    MKL_LDFLAGS += $(RPATH_MKL)
  endif
endif
LIBBLAS   := $(MKL_LDFLAGS)
LIBLAPACK := $(MKL_LDFLAGS)
endif
```

This seems problematic because `$(MKLROOT)/lib/intel64`  doesn’t exist:
```
$ echo $MKLROOT/lib/intel64
/opt/intel/compilers_and_libraries_2017.4.181/mac/mkl/lib/intel64
$ ls $MKLROOT/lib/intel64
ls: cannot access /opt/intel/compilers_and_libraries_2017.4.181/mac/mkl/lib/intel64: No such file or directory
```

However, `MKLROOT/lib` *does* have libmkl_rt.dylib!
```
$ file $MKLROOT/lib/libmkl_rt.dylib
/opt/intel/compilers_and_libraries_2017.4.181/mac/mkl/lib/libmkl_rt.dylib: Mach-O universal binary with 2 architectures: [i386: Mach-O dynamically linked shared library i386] [x86_64]
/opt/intel/compilers_and_libraries_2017.4.181/mac/mkl/lib/libmkl_rt.dylib (for architecture i386):	Mach-O dynamically linked shared library i386
/opt/intel/compilers_and_libraries_2017.4.181/mac/mkl/lib/libmkl_rt.dylib (for architecture x86_64):	Mach-O 64-bit dynamically linked shared library x86_64
```

I modify both lines that set `MKLLIB` to be: `MKLLIB := $(MKLROOT)/lib`, then rerun `make`.

I get the same error.

Questions:
- should I be concerned that `RPATH` is being modified in the snippet above only for Linux and not Darwin?
- Is there any way to see what the `vendor()` function in `base/linalg/blas.jl` is using for `Base.libblas_name`? Or just more information about what Julia is trying to load?

I’m confident my MKL setup is correct, I’ve build libraries and code using them, and I’d like to get to the bottom of this, so any tips on how to better troubleshoot Make/Julia will be most appreciated!

**Comment by fasiha** [comment]
Oh, ok: I collapsed the excerpt from Make.inc quoted in my previous comment (starting at line 971) to the following, and it works:
```make
export MKL_INTERFACE_LAYER := ILP64
MKLLIB := $(MKLROOT)/lib
USE_SYSTEM_BLAS:=1
USE_SYSTEM_LAPACK:=1
LIBBLASNAME := libmkl_rt
LIBLAPACKNAME := libmkl_rt
MKL_LDFLAGS := -L$(MKLLIB) -lmkl_rt
RPATH_MKL := -Wl,-rpath,$(MKLLIB)
RPATH     += $(RPATH_MKL)
MKL_LDFLAGS += $(RPATH_MKL)
LIBBLAS   := $(MKL_LDFLAGS)
LIBLAPACK := $(MKL_LDFLAGS)
```
I.e., I removed anything about WINNT and kept the snippet about Linux.

(I can indirectly confirm that MKL is working: 
```julia
x = reshape(1:128 * 128 * 128, 128, 128, 128);
@elapsed y = fft(x)
@elapsed fft!(y)
```
settles to 0.025852694 & 0.00725526 seconds after warmup with the above build. Meanwhile, official Julia 0.5.2 macOS build converges to 0.06400 and 0.041165714 seconds for the two times. Matlab for the equivalent case, `x = reshape(1:128^3, 128, 128, 128); timeit(@() fftn(x))`, converges to 0.020675 seconds, within 3% of Julia tentatively with MKL.)

Can someone help me debug the Makefile and make it macOS-friendly? 🙇
